File: lakeside-mutual/customer-self-service-backend-constraints/src/main/java/com/lakesidemutual/customerselfservice/interfaces/PolicyCreationExpirationMessageConsumer.java
package com.lakesidemutual.customerselfservice.interfaces;

import com.lakesidemutual.customerselfservice.domain.insurancequoterequest.InsuranceQuoteEvent;
import com.lakesidemutual.customerselfservice.domain.insurancequoterequest.InsuranceQuoteExpiredEvent;
import com.lakesidemutual.customerselfservice.domain.insurancequoterequest.InsuranceQuoteRequestAggregateRoot;
import com.lakesidemutual.customerselfservice.domain.insurancequoterequest.PolicyCreatedEvent;
import com.lakesidemutual.customerselfservice.infrastructure.InsuranceQuoteRequestRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.*;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.groupingBy;

/**
 * PolicyCreationExpirationMessageConsumer is a Spring component that consumes PolicyCreatedEvents
 * and InsuranceQuoteExpiredEvents as they arrive through the Kafka topic.
 * It processes these events by updating the status of the corresponding insurance quote requests.
 * */
@Component
public class PolicyCreationExpirationMessageConsumer {
	private final Logger logger = LoggerFactory.getLogger(this.getClass());

	List<InsuranceQuoteEvent> queue = new LinkedList<>();

	@Autowired
	private InsuranceQuoteRequestRepository insuranceQuoteRequestRepository;

	@KafkaListener(topics = {"${policyCreatedEvent.topicName}", "${insuranceQuoteExpiredEvent.topicName}"},
			groupId = "${spring.kafka.consumer.group-id}",
			containerFactory = "policyCreationExpirationListenerFactory")
	public void receiveEvent(final InsuranceQuoteEvent event) {
		queue.add(event);
	}

	@Scheduled(fixedRate = 1000, initialDelay = 10000)
	private void filterAndHandleEvents() {
		logger.info("Insurance events publish initiated with " + this.queue.size() + " elements.");

		List<InsuranceQuoteEvent> buffered = this.queue;
		this.queue = new LinkedList<>();

		Map<Long, List<InsuranceQuoteEvent>> eventsPerInsuranceQuote = buffered
			.stream()
			.collect(groupingBy(InsuranceQuoteEvent::getInsuranceQuoteRequestId));

		Map<Long, List<InsuranceQuoteEvent>> filteredEventsPerInsuranceQuote = eventsPerInsuranceQuote
			.entrySet()
			.stream()
			.collect(Collectors.toMap(Map.Entry::getKey, e -> filterInsuranceQuoteEvents(e.getValue())));

		filteredEventsPerInsuranceQuote
			.forEach((k, v) -> v.forEach(this::handleEvent));
	}

	private void handleEvent(final InsuranceQuoteEvent event) {

		if (event instanceof InsuranceQuoteExpiredEvent) {
			InsuranceQuoteExpiredEvent insuranceQuoteExpiredEvent = (InsuranceQuoteExpiredEvent) event;

			String line = event.getInsuranceQuoteRequestId().toString() + " " +
					insuranceQuoteExpiredEvent.getDate().getTime() + " "
					+ System.currentTimeMillis() + " InsuranceQuoteExpiredEvent\n";
			try {
				Files.write(Paths.get("received-insurance-events.txt"), line.getBytes(), StandardOpenOption.APPEND);
			} catch (IOException e) {}

			//insuranceQuoteRequest.markQuoteAsExpired(insuranceQuoteExpiredEvent.getDate());
			//logger.info("The insurance quote for insurance quote request " + insuranceQuoteExpiredEvent.getInsuranceQuoteRequestId() + " has expired.");
		} else if (event instanceof PolicyCreatedEvent) {
			PolicyCreatedEvent policyCreatedEvent = (PolicyCreatedEvent) event;

			String line = event.getInsuranceQuoteRequestId().toString() + " " +
					policyCreatedEvent.getDate().getTime() + " "
					+ System.currentTimeMillis() + " PolicyCreatedEvent\n";
			try {
				Files.write(Paths.get("received-insurance-events.txt"), line.getBytes(), StandardOpenOption.APPEND);
			} catch (IOException e) {}

			//insuranceQuoteRequest.finalizeQuote(policyCreatedEvent.getPolicyId(), policyCreatedEvent.getDate());
			//logger.info("The policy for for insurance quote request " + policyCreatedEvent.getInsuranceQuoteRequestId() + " has been created.");
		}

		//insuranceQuoteRequestRepository.save(insuranceQuoteRequest);
	}

	private List<InsuranceQuoteEvent> filterInsuranceQuoteEvents(List<InsuranceQuoteEvent> events) {
		if (events.size() > 1 && events.get(0) instanceof InsuranceQuoteExpiredEvent) {
			return events.subList(1, 2);
		}
		return events;
	}
}

File: lakeside-mutual/customer-self-service-backend-integrated/src/main/scala/com/lakesidemutual/customerselfservice/constraints/PolicyCreationExpirationMessageConsumer.scala
package com.lakesidemutual.customerselfservice.constraints

import java.nio.file.{Files, Paths, StandardOpenOption}
import java.time.Duration
import java.util.Properties

import com.github.annterina.stream_constraints.CStreamsBuilder
import com.github.annterina.stream_constraints.constraints.ConstraintBuilder
import com.github.annterina.stream_constraints.constraints.window.WindowConstraintBuilder
import com.lakesidemutual.customerselfservice.constraints.serde.InsuranceQuoteEventSerde
import com.lakesidemutual.customerselfservice.domain.insurancequoterequest.{InsuranceQuoteEvent, InsuranceQuoteExpiredEvent, PolicyCreatedEvent}
import com.lakesidemutual.customerselfservice.infrastructure.InsuranceQuoteRequestRepository
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.common.serialization.{Serde, Serdes}
import org.apache.kafka.streams.{KafkaStreams, StreamsConfig, Topology}
import org.apache.kafka.streams.scala.kstream.Consumed
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.{Bean, Configuration}

@Configuration
class PolicyCreationExpirationMessageConsumer {
  val logger: Logger = LoggerFactory.getLogger(this.getClass)

  val KAFKA_BOOTSTRAP_SERVERS = "localhost:9092"
  val STREAMS_APP_NAME = "policy-creation-expiration-streams"
  val INSURANCE_QUOTE_EXPIRED_TOPIC_NAME = "insurance-quote-expired-events"
  val POLICY_CREATED_TOPIC_NAME = "policy-created-events"

  val insuranceQuoteEventSerde: Serde[InsuranceQuoteEvent] = Serdes.serdeFrom(InsuranceQuoteEventSerde.serializer(),
    InsuranceQuoteEventSerde.deserializer())

  @Autowired
  var insuranceQuoteRequestRepository: InsuranceQuoteRequestRepository = null

  @Bean
  def kafkaStreams(): KafkaStreams = {
    val kafkaStreamsConfig: Properties = {
      val properties = new Properties()
      properties.put(StreamsConfig.APPLICATION_ID_CONFIG, "policy-expiration-constraints-application")
      properties.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092")
      properties.put(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG, "DEBUG")
      properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest")
      properties
    }

    val insuranceQuoteEventSerde = Serdes.serdeFrom(InsuranceQuoteEventSerde.serializer(),
      InsuranceQuoteEventSerde.deserializer())

    val windowConstraint = new WindowConstraintBuilder[String, InsuranceQuoteEvent]
      .before(((_, e) => e.isInstanceOf[InsuranceQuoteExpiredEvent], "insurance-quote-expired"))
      .after(((_, e) => e.isInstanceOf[PolicyCreatedEvent], "policy-created"))
      .window(Duration.ofMillis(2000))
      .dropBefore

    val constraint = new ConstraintBuilder[String, InsuranceQuoteEvent, java.lang.Long]
      .windowConstraint(windowConstraint)
      .link((_, e) => e.getInsuranceQuoteRequestId)(Serdes.Long)
      .build(Serdes.String, insuranceQuoteEventSerde)

    val builder = new CStreamsBuilder()

    builder
      .stream(Set("insurance-quote-expired-events", "policy-created-events"))(Consumed.`with`(Serdes.String, insuranceQuoteEventSerde))
      .constrain(constraint)
      .foreach((_, event) => handleEvent(event))

    val topology: Topology = builder.build()

    val streams = new KafkaStreams(topology, kafkaStreamsConfig)

    streams.start()
    streams
  }

  private def handleEvent(event: InsuranceQuoteEvent): Unit =  {
//  		val id = event.getInsuranceQuoteRequestId
//  		val insuranceQuoteRequestOpt = insuranceQuoteRequestRepository.findById(id);
//
//  		if(!insuranceQuoteRequestOpt.isPresent) {
//  			logger.error("Unable to process the event with an invalid insurance quote request id.")
//  			return
//  		}
//
//  		val insuranceQuoteRequest = insuranceQuoteRequestOpt.get()

  		event match {
        case insuranceQuoteExpiredEvent: InsuranceQuoteExpiredEvent =>
          logger.info("A new insurance expiration event has been received.")

          val line = insuranceQuoteExpiredEvent.getInsuranceQuoteRequestId.toString + " " +
            insuranceQuoteExpiredEvent.getDate.getTime + " " + System.currentTimeMillis() + " InsuranceQuoteExpiredEvent\n"

          Files.write(Paths.get("received-insurance-events-sc2000.txt"), line.getBytes, StandardOpenOption.APPEND)

        //insuranceQuoteRequest.markQuoteAsExpired(insuranceQuoteExpiredEvent.getDate)
        //logger.info("The insurance quote for insurance quote request " + insuranceQuoteRequest.getId + " has expired.");

        case policyCreatedEvent:  PolicyCreatedEvent =>
          logger.info("A new policy creation event has been received.")

          val line = policyCreatedEvent.getInsuranceQuoteRequestId.toString + " " +
            policyCreatedEvent.getDate.getTime + " " + System.currentTimeMillis() + " PolicyCreatedEvent\n"

          Files.write(Paths.get("received-insurance-events-sc2000.txt"), line.getBytes, StandardOpenOption.APPEND)

          //insuranceQuoteRequest.finalizeQuote(policyCreatedEvent.getPolicyId, policyCreatedEvent.getDate)
          //logger.info("The policy for for insurance quote request " + insuranceQuoteRequest.getId + " has been created.");
      }

  		//insuranceQuoteRequestRepository.save(insuranceQuoteRequest);
  	}
}
